pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +237,306 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
+		commitRecTs = fallbackMillis
+	}
+
+	level.Debug(logger).Log(
+		"msg", "creating partition state",
+		"partition", lag.Partition,
+		"lag", lag.Lag,
+		"commit_rec_ts", commitRecTs,
+		"commit_rec_offset", lag.Commit.At,
+		"last_seen_offset", lastSeenOffset,
+		"last_block_end_ts", lastBlockEndTs,
+	)
+
+	return partitionState{
+		Lag:                   lag.Lag,
+		Commit:                lag.Commit,
+		CommitRecordTimestamp: time.UnixMilli(commitRecTs),
+		LastSeenOffset:        lastSeenOffset,
+		LastBlockEnd:          time.UnixMilli(lastBlockEndTs),
+	}
+}
+
+// consumePartition consumes records from the given partition until the cycleEnd timestamp.
+// If the partition is lagging behind, it takes care of calling consuming it in sections.
+func (b *BlockBuilder) consumePartition(ctx context.Context, partition int32, state partitionState, cycleEnd time.Time) (err error) {
+	builder := NewTSDBBuilder(b.logger, b.cfg.DataDir, b.cfg.BlocksStorage, b.limits, b.tsdbBuilderMetrics)
+	defer runutil.CloseWithErrCapture(&err, builder, "closing tsdb builder")
+
+	// Section is a portion of the partition to process in a single pass. One cycle may process multiple sections if the partition is lagging.
+	sectionEnd := cycleEnd
+	if sectionEnd.Sub(state.CommitRecordTimestamp) > time.Duration(1.5*float64(b.cfg.ConsumeInterval)) {
+		// We are lagging behind by more than 1.5*interval or there is no commit. We need to consume the partition in sections.
+		// We iterate through all the ConsumeInterval intervals, starting from the first one after the last commit until the cycleEnd,
+		// i.e. [T, T+interval), [T+interval, T+2*interval), ... [T+S*interval, cycleEnd)
+		// where T is the CommitRecordTimestamp, the timestamp of the record, whose offset we committed previously.
+		// When there is no kafka commit, we play safe and assume LastSeenOffset, and LastBlockEnd were 0 to not discard any samples unnecessarily.
+		sectionEnd, _ = nextCycleEnd(
+			state.CommitRecordTimestamp,
+			b.cfg.ConsumeInterval,
+			b.cfg.ConsumeIntervalBuffer,
+		)
+		level.Info(b.logger).Log("msg", "partition is lagging behind the cycle", "partition", partition, "lag", state.Lag, "section_cycle_end", sectionEnd, "cycle_end", cycleEnd, "commit_rec_ts", state.CommitRecordTimestamp)
+	}
+	for !sectionEnd.After(cycleEnd) {
+		partitionLogger := log.With(b.logger, "partition", partition, "lag", state.Lag, "section_cycle_end", sectionEnd)
+		state, err = b.consumePartitionSection(ctx, partitionLogger, builder, partition, state, sectionEnd)
+		if err != nil {
+			return fmt.Errorf("consume partition %d: %w", partition, err)
+		}
+		sectionEnd = sectionEnd.Add(b.cfg.ConsumeInterval)
+	}
+
+	return nil
+}
+
+func (b *BlockBuilder) consumePartitionSection(ctx context.Context, logger log.Logger, builder *TSDBBuilder, partition int32, state partitionState, sectionEnd time.Time) (_ partitionState, retErr error) {
```

Comment:
I went with Vladimir through different cases to better understand how the offset work and LGTM. At a first glance I found it non trivial to understand, so I'm leaving my notes here.

Basic info:
- Next consumption cycle should process `partitionState.Commit` offset as 1st record.


Partition has 1 record, and the record has been partially processed:

- AddConsumePartitions() called with `state.Commit.At`
  - The first record will have offset `state.Commit.At`
- firstRec  = first record (not fully processed)
  - It has offset `state.Commit.At`
- commitRec = Offset set to "state.Commit.At - 1"
- lastRec = firstRec
- We commit `commit.At` set to `commitRec.Offset + 1`, which is `state.Commit.At - 1 + 1`, which is `state.Commit.At`, correct


Partition has 2 records, the 1st record is fully processed, but the 2nd record is partially processed:

- AddConsumePartitions() called with `state.Commit.At`
  - The first record will have offset `state.Commit.At`
- firstRec  = first record (fully processed)
  - It has offset `state.Commit.At`
- The 2nd record is partially processed
  - It has offset `state.Commit.At + 1`
- commitRec = firstRecord
  - It has offset `state.Commit.At`
- We commit `commit.At` set to `commitRec.Offset + 1`, which is `state.Commit.At + 1`, which is the 2nd record, which is the earlier record with samples not fully processed and so the next time we'll have to restart from there, correct


Partition has 2 records, we fully process all records:

- AddConsumePartitions() called with `state.Commit.At`
  - The first record will have offset `state.Commit.At`
- firstRec  = first record (fully processed)
  - It has offset `state.Commit.At`
- The 2nd record (fully processed)
  - It has offset `state.Commit.At + 1`
- commitRec = second record
  - It has offset `state.Commit.At + 1`
- We commit `commit.At` set to `commitRec.Offset + 1`, which is `state.Commit.At + 1 + 1`, which is the offset that will have the 3rd record, which is correct because records up until the 2nd one (included) have been fully processed, correct


Partition has 1 record, it's first record ever in the partition so its offset is 0 and the record is partially processed:

- AddConsumePartitions() called with `state.Commit.At = 0`
  - The first record will have offset `state.Commit.At = 0`
- firstRec  = first record (partially processed)
  - It has offset `state.Commit.At = 0`
- commitRec = offset set to `state.Commit.At - 1`, which is `-1`
- We commit `commit.At` set to `commitRec.Offset + 1`, which is `-1 +1`, which is `0`, correct